### PR TITLE
Fix incorrect parsing of bfrange (#631)

### LINKED ISSUE #631: Allowed memory exhausted when parse the PDF file.
- PHP Version: 7.4.33
 - PDFParser Version: 2.5.0

### Description:
Trying to parse this PDF always result in Allowed memory exhausted error. 

Error: Allowed memory size of 1077936128 bytes exhausted (tried to
allocate 335544320 bytes) in
...../smalot/pdfparser/src/Smalot/PdfParser/Font.php,
line 223

Set up PHP memory limit to 4GB did not help either.
I have also tried to setDecodeMemoryLimit to lower but still had the same memory issue. Setting Decode memory limit prevent the error only when I set it to 1000 or lower. So maybe it should be set in MB and not in bytes, or there is an bug in the code.
 

### PDF input

[test_pdf.pdf](https://github.com/smalot/pdfparser/files/12291120/test_pdf.pdf)

### Expected output & actual output

Parser should either parse the text from the PDF, or return empty string or some exception and not memory error.

### Code

```
$config = new Config();
$url = 'path_to_PDF_folder/test_pdf.pdf';
$config->setRetainImageContent(false);
$config->setDecodeMemoryLimit(1000000);
$parser = new Parser([], $config);
$pdf = $parser->parseFile($url);
```

### PR DESCRIPTION
**Pull request by nsfisis**
# Type of pull request

* [x] Bug fix (involves code and configuration changes)
* [ ] New feature (involves code and configuration changes)
* [ ] Documentation update
* [ ] Something else

# About

fixes #631

Previously, the regular expression for single-offset bfrange mapping unintentionally matched portions of the bracketed array form. This caused extremely large code range, leading to out of memory.

Now, we capture the `<from>` and `<to>` ranges first and then distinguish between single-offset (`<xxxx>`) and array (`[<xxxx> <xxxx> ...]`) forms. This ensures that the single-offset regex won't accidentally match bracketed segments.

# Checklist for code / configuration changes

See [CONTRIBUTING.md](./../CONTRIBUTING.md) for all essential information about contributing.

# TODO

- [x] Write tests.

Question:
In order to write test code, I need to generate a PDF that contains arbitrary `bfrange` sections. However, I’m afraid I don’t know how to do that. Do you have any suggestions on how to create such a file?

### CODE REVIEW

**Comment by k00ni** [comment]
> In order to write test code, I need to generate a PDF that contains arbitrary bfrange sections. However, I’m afraid I don’t know how to do that. Do you have any suggestions on how to create such a file?

You don't need a real PDF file, instead create a `Font` instances and set them up depending on your test case. If you want to test how it handles `bfrange` sections, you can create different strings with correct and faulty data and inject it in the instance. You might need to use other class instances too to build a working case, but it seems reasonable and doable.

Here are a few starting points which might help:
* https://github.com/smalot/pdfparser/blob/master/tests/PHPUnit/Unit/FontTest.php
* https://github.com/smalot/pdfparser/blob/master/tests/PHPUnit/Integration/FontTest.php#L165
* https://github.com/smalot/pdfparser/blob/master/tests/PHPUnit/Integration/PagesTest.php

**Comment by nsfisis** [comment]
Added tests: `tests/PHPUnit/Integration/FontTest.php`

Ready for review!

File: src/Smalot/PdfParser/Font.php [hunk 1]
```diff
@@ -216,45 +216,46 @@ public function loadTranslateTable(): array
             // Support for multiple bfrange sections
             if (preg_match_all('/beginbfrange(?P<sections>.*?)endbfrange/s', $content, $matches)) {
                 foreach ($matches['sections'] as $section) {
-                    // Support for : <srcCode1> <srcCode2> <dstString>
-                    $regexp = '/<(?P<from>[0-9A-F]+)> *<(?P<to>[0-9A-F]+)> *<(?P<offset>[0-9A-F]+)>[ \r\n]+/is';
+                    // Regexp to capture <from>, <to>, and either <offset> or [...] items.
+                    // - (?P<from>...) Source range's start
+                    // - (?P<to>...)   Source range's end
+                    // - (?P<dest>...) Destination range's offset or each char code
+                    //                 Some PDF file has 2-byte Unicode values on new lines > added \r\n
```
Comment (k00ni) [inline src/Smalot/PdfParser/Font.php:223]:
```suggestion
                    /**
                     * Regexp to capture <from>, <to>, and either <offset> or [...] items.
                     * - (?P<from>...) Source range's start
                     * - (?P<to>...)   Source range's end
                     *  - (?P<dest>...) Destination range's offset or each char code
                     *                 Some PDF file has 2-byte Unicode values on new lines > added \r\n
                     */
```

File: src/Smalot/PdfParser/Font.php [hunk 2]
```diff
@@ -216,45 +216,46 @@ public function loadTranslateTable(): array
...
                         $char_from = hexdec($from);
                         $char_to = hexdec($matches['to'][$key]);
-                        $offset = hexdec($matches['offset'][$key]);
+                        $dest = $matches['dest'][$key];
 
-                        for ($char = $char_from; $char <= $char_to; ++$char) {
-                            $this->table[$char] = self::uchr($char - $char_from + $offset);
-                        }
-                    }
+                        if (preg_match('/^<(?P<offset>[0-9A-F]+)>$/i', $dest, $offset_matches)) {
```
Comment (k00ni) [inline src/Smalot/PdfParser/Font.php:233]:
```suggestion
                        if (1 === preg_match('/^<(?P<offset>[0-9A-F]+)>$/i', $dest, $offset_matches)) {
```

File: src/Smalot/PdfParser/Font.php [hunk 3]
```diff
@@ -216,45 +216,46 @@ public function loadTranslateTable(): array
...
-                            );
-                            $text = '';
-                            foreach ($parts as $part) {
-                                $text .= self::uchr(hexdec($part));
+                        } else {
+                            // Support for : <srcCode1> <srcCodeN> [<dstString1> <dstString2> ... <dstStringN>]
+                            $strings = [];
+                            preg_match_all('/<(?P<string>[0-9A-F]+)> */is', $dest, $strings);
+
+                            foreach ($strings['string'] as $position => $string) {
```
Comment (k00ni) [inline src/Smalot/PdfParser/Font.php:246]: Can we be sure that `$strings['string']` key is always set?

File: src/Smalot/PdfParser/Font.php [hunk 4]
```diff
@@ -216,45 +216,46 @@ public function loadTranslateTable(): array
...
+                                $parts = preg_split(
+                                    '/([0-9A-F]{4})/i',
+                                    $string,
+                                    0,
+                                    \PREG_SPLIT_NO_EMPTY | \PREG_SPLIT_DELIM_CAPTURE
+                                );
+                                $text = '';
+                                foreach ($parts as $part) {
```
Comment (k00ni) [inline src/Smalot/PdfParser/Font.php:254]: In case `preg_split` returns `false` PHP will complain. Maybe add an if-clause before the foreach-loop to check if `$parts` is valid?

File: tests/PHPUnit/Integration/FontTest.php [hunk 5]
```diff
@@ -238,6 +238,68 @@ public function testLoadTranslateTable(): void
         $this->assertEquals('y', $table[92]);
     }
 
+    /**
+     * Tests loadTranslateTable with a bfrange definition that lists every destination character.
+     *
+     * @see https://github.com/smalot/pdfparser/issues/631
+     */
+    public function testLoadTranslateTableIssue631(): void
```
Comment (k00ni) [inline tests/PHPUnit/Integration/FontTest.php:246]: Thanks for investing the time to add a test!

This seems to test the handling of elements, which couldn't be handled before. Would you mind elaborate a bit more on the this test? Is it worth thinking about further tests for other (edge) cases?

**Review by k00ni — CHANGES_REQUESTED** [review]
First of all, thanks for your PR. Sorry for the delayed feedback, I am super busy currently.

There are only a few small remarks and suggestions. I hope you have some time left so we can finish this together. Looking forward to your repsonse.

**Comment by k00ni** [comment]
_Just for future reference:_ info about bfrange handling in different PDF specifications. Because we support different versions it might be interesting what was defined and when.

- **PDF Specification 1.3**: see page 326 and later (https://opensource.adobe.com/dc-acrobat-sdk-docs/pdfstandards/pdfreference1.3.pdf)
- **PDF Specification 2.0**: see page 349 and later (https://developer.adobe.com/document-services/docs/assets/5b15559b96303194340b99820d3a70fa/PDF_ISO_32000-2.pdf)

File: src/Smalot/PdfParser/Font.php (hunk 3 above)
Comment (nsfisis) [reply src/Smalot/PdfParser/Font.php:250]: I added check if `preg_match_all()` succeeds.

File: src/Smalot/PdfParser/Font.php (hunk 4 above)
Comment (nsfisis) [reply src/Smalot/PdfParser/Font.php:261]: I added check if `preg_split()` succeeds.

File: tests/PHPUnit/Integration/FontTest.php (hunk 5 above)
Comment (nsfisis) [reply tests/PHPUnit/Integration/FontTest.php:246]: I added more valid or invalid test cases.

File: src/Smalot/PdfParser/Font.php [hunk 6]
```diff
@@ -216,45 +216,53 @@ public function loadTranslateTable(): array
             // Support for multiple bfrange sections
             if (preg_match_all('/beginbfrange(?P<sections>.*?)endbfrange/s', $content, $matches)) {
                 foreach ($matches['sections'] as $section) {
-                    // Support for : <srcCode1> <srcCode2> <dstString>
-                    $regexp = '/<(?P<from>[0-9A-F]+)> *<(?P<to>[0-9A-F]+)> *<(?P<offset>[0-9A-F]+)>[ \r\n]+/is';
+                    /**
+                     * Regexp to capture <from>, <to>, and either <offset> or [...] items.
+                     * - (?P<from>...) Source range's start
+                     * - (?P<to>...)   Source range's end
+                     *  - (?P<dest>...) Destination range's offset or each char code
```
Comment (k00ni) [inline src/Smalot/PdfParser/Font.php:223]:
```suggestion
                     * - (?P<dest>...) Destination range's offset or each char code
```

**Comment by k00ni** [comment]
Thank you @nsfisis
